TaskManager/server.js

```diff
@@ -0,0 +1,147 @@
+
+// Request handlers
+
+	// Return all tasks (fetch request in client.js l.107)
+app.get(routes.tasks, (req, res) => {
+	res.json(getAllTasks())
+})
+
+	// Create new task (fetch request in client.js l.149)
+app.post(routes.tasks, (req, res) => {
+	const newTask = req.body;
+	const allTasks = getAllTasks(); 
+	allTasks.tasks.push(newTask); 
+	updateAllTasks(allTasks); 
+	res.send();
+})
+
+	// Mark as completed (fetch request in client.js l.205)
+app.post(routes.finishTask, (req, res) => {
+	const taskIndex = req.body.taskIndex;
+	const allTasks = getAllTasks();
+	const targetTask = allTasks.tasks.splice(taskIndex, 1)[0];
+	targetTask.completedOn = Date.now();
+	allTasks.completedTasks.push(targetTask);
+	updateAllTasks(allTasks);
+	res.send();
+})
+
+	// Update task (fetch request in client.js l.263)
+app.put(routes.tasks, (req, res) => {
+	const updatedTask = req.body.updatedTask;
+	const taskIndex = req.body.taskIndex;
+	const allTasks = getAllTasks();
+	allTasks.tasks[taskIndex] = updatedTask;
+	updateAllTasks(allTasks);
+	res.send();
+})
+
+	// Delete task (fetch request in client.js: tasks l.178, completed tasks l.229)
+app.delete(routes.tasks, (req, res) => {
+	const taskIndex = req.body.taskIndex;
+	const isCompleted = req.body.isCompleted;
+	const allTasks = getAllTasks();
+	if (isCompleted) {
+		allTasks.completedTasks.splice(taskIndex, 1);
+	} else {
+		allTasks.tasks.splice(taskIndex, 1);
+	}
+	updateAllTasks(allTasks);
+	res.send();
+})
+
+	// Login POST (fetch request in login.js l.29)
+app.post(routes.login, (req, res) => {
+
+	const username = req.body.username;
+	const password = req.body.password;
+	const auth = {
+		success: false,
+		role: null,
+		msg: 'Invalid credentials. Please try again.',
+	};
+
+	users.find((user) => {
+		if (user.username === username && user.password === password) {
+			auth.success = true;
+			auth.role = user.role;
+			auth.msg = `Welcome, ${username}! You are logged in as ${user.role}.`;
+		}
+	});
+	console.log(`Authentication is ${auth.success}`);
+	console.log(auth.msg);
+	res.json(auth);
+
+	// FIX - code below disfunctional, implemented on clientside instead (login.js l.40).
+	// if (auth.success) {
+	// 	console.log(`Authentication is ${auth.success}`);
+	// 	console.log(auth.msg);
+	// 	res.redirect(routes.index);
+	// } else {
+	// 	res.redirect(routes.login);
+	// };
+});
+
+	// Root URL (no fetch request or link in client)
+app.get(routes.login, (req, res) => {
+	res.sendFile(__dirname + '/client/login.html');
+});
+
+	// Redirect (no fetch request or link in client)
+app.get(routes.index, (req, res) => {
+	res.sendFile(__dirname + '/client/viewtasks.html');
+});
+
+	// Signup POST (fetch request in signup.js l.29)
+app.post(routes.signup, (req, res) => {
+	const username = req.body.username;
+	const password = req.body.password;
+	const user = {
+		username: username,
+		password: password,
+		role: 'user',
+	};
+	const users = getUsers();
+	users.push(user);
+	updateUsers(users);
```

Comment:
```suggestion
	updateUsers([...users, user]);
```
Another shorthand writing you can employ :) Can read about the [Spread Syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax)